### ISSUE TITLE
[Question]: How to set data loader when training with PP and DP

### ISSUE THREAD

**Issue by DarrenYing** [Oneflow-Inc/oneflow#10039]
### Description

我在一个四卡的机器上运行流水线并行+数据并行的训练，其中gpu0和gpu1是数据并行关系，负责Stage0，gpu2和gpu3也是数据并行关系，负责Stage1
数据加载相关代码如下：
```python
S0 = flow.sbp.split(dim=0)
P01 = flow.placement("cuda", ranks=[0, 1])
P23 = flow.placement("cuda", ranks=[2, 3])
P0123 = flow.placement("cuda", ranks=[0, 1, 2, 3])

train_dataset = FakeDataset()

sampler = data.DistributedSampler(train_dataset, drop_last=True)

train_dataloader = flow.utils.data.DataLoader(
    dataset=train_dataset,
    batch_size=1,
    sampler=sampler,
    drop_last=True,
)

#  other code
...

# train
for batch_idx, batch_data in enumerate(train_dataloader, 1):
    batch_data = batch_data.to_global(placement=P0123, sbp=S0).to_global(placement=P01, sbp=S0)
    ...
```
按这种写法，`global-batch-size`为4，gpu0和gpu1上各分了`batch-size=2`

我希望，使`global-batch-size`为2，gpu0和gpu1上各分了`batch-size=1`，做了如下修改：
```python
train_dataset = FakeDataset()

rank = flow.env.get_rank()

if rank in [0, 1]:
    sampler = data.distributed.DistributedSampler(train_dataset, drop_last=True, num_replicas=2)
else:
    sampler = data.distributed.DistributedSampler(train_dataset, rank=rank-2, num_replicas=2)

train_dataloader = flow.utils.data.DataLoader(
    dataset=train_dataset,
    batch_size=1,
    sampler=sampler,
    drop_last=True,
)

#  other code
...

# train
for batch_idx, batch_data in enumerate(train_dataloader, 1):
    batch_data = batch_data.to_global(placement=P01, sbp=S0)
    ...
```
我开启了`graph.debug(1)`，并查看显存占用，发现实现了`global-batch-size`为2，gpu0和gpu1上各分了`batch-size=1`，但是实际训练过程中，我发现模型收敛速度变慢，这里的变慢是指，和`global-batch-size`为2的`PP4`模型对比，和`global-batch-size`为4的`PP2+DP2`模型（即上面给出的第一种写法）对比，都变慢了，而且相差程度较大，应该不是`shuffle`导致的。

想请教一下，是不是我这种修改方式存在问题，正确的修改方式应该是什么样的呢？十分感谢！

**Comment by Ldpe2G** [comment]
```bash
train_dataset = FakeDataset()
local_rank = flow.env.get_rank()

S0 = flow.sbp.split(dim=0)
P01 = flow.placement("cuda", ranks=[0, 1])
P23 = flow.placement("cuda", ranks=[2, 3])

sampler = data.distributed.DistributedSampler(train_dataset, drop_last=True, num_replicas=2, rank=(local_rank  % 2))

train_dataloader = flow.utils.data.DataLoader(
    dataset=train_dataset,
    batch_size=1,
    sampler=sampler,
    drop_last=True,
)

#  other code
...

# train
for batch_idx, batch_data, batch_label in enumerate(train_dataloader, 1):
    batch_data = batch_data.to_global(placement=P01, sbp=S0)
    batch_label = batch_label.to_global(placement=P01, sbp=S0).to_global(placement=P23)


```

应该是要这样改，对于 dataloader  rank 0,1 和 rank 2,3 分别为单独的数据并行组，每个组都是读取完整的数据集。然后因为是流水并行，数据和label都只保留 rank0和1。
然后Label需要额外一步，将 stage 0 上的保留的数据，to 到 stage 1 上，这样数据和label才能对的上。

**Comment by DarrenYing** [comment]
> ```shell
> train_dataset = FakeDataset()
> local_rank = flow.env.get_rank()
> 
> S0 = flow.sbp.split(dim=0)
> P01 = flow.placement("cuda", ranks=[0, 1])
> P23 = flow.placement("cuda", ranks=[2, 3])
> 
> sampler = data.distributed.DistributedSampler(train_dataset, drop_last=True, num_replicas=2, rank=(local_rank  % 2))
> 
> train_dataloader = flow.utils.data.DataLoader(
>     dataset=train_dataset,
>     batch_size=1,
>     sampler=sampler,
>     drop_last=True,
> )
> 
> #  other code
> ...
> 
> # train
> for batch_idx, batch_data, batch_label in enumerate(train_dataloader, 1):
>     batch_data = batch_data.to_global(placement=P01, sbp=S0)
>     batch_label = batch_label.to_global(placement=P01, sbp=S0).to_global(placement=P23)
> ```
> 
> 应该是要这样改，对于 dataloader rank 0,1 和 rank 2,3 分别为单独的数据并行组，每个组都是读取完整的数据集。然后因为是流水并行，数据和label都只保留 rank0和1。 然后Label需要额外一步，将 stage 0 上的保留的数据，to 到 stage 1 上，这样数据和label才能对的上。

感觉
```python
sampler = data.distributed.DistributedSampler(train_dataset, drop_last=True, num_replicas=2, rank=(local_rank  % 2))
```
和我原来的写法好像没有本质区别
```python
if rank in [0, 1]:
    sampler = data.distributed.DistributedSampler(train_dataset, drop_last=True, num_replicas=2)
else:
    sampler = data.distributed.DistributedSampler(train_dataset, rank=rank-2, num_replicas=2)
```
谢谢，我会去试试，看看会不会不一样

我的label就是batch_data本身，stage1 上的预测结果我会 to 到 stage0 上，再计算loss的，应该没有问题

**Comment by DarrenYing** [comment]
> ```shell
> train_dataset = FakeDataset()
> local_rank = flow.env.get_rank()
> 
> S0 = flow.sbp.split(dim=0)
> P01 = flow.placement("cuda", ranks=[0, 1])
> P23 = flow.placement("cuda", ranks=[2, 3])
> 
> sampler = data.distributed.DistributedSampler(train_dataset, drop_last=True, num_replicas=2, rank=(local_rank  % 2))
> 
> train_dataloader = flow.utils.data.DataLoader(
>     dataset=train_dataset,
>     batch_size=1,
>     sampler=sampler,
>     drop_last=True,
> )
> 
> #  other code
> ...
> 
> # train
> for batch_idx, batch_data, batch_label in enumerate(train_dataloader, 1):
>     batch_data = batch_data.to_global(placement=P01, sbp=S0)
>     batch_label = batch_label.to_global(placement=P01, sbp=S0).to_global(placement=P23)
> ```
> 
> 应该是要这样改，对于 dataloader rank 0,1 和 rank 2,3 分别为单独的数据并行组，每个组都是读取完整的数据集。然后因为是流水并行，数据和label都只保留 rank0和1。 然后Label需要额外一步，将 stage 0 上的保留的数据，to 到 stage 1 上，这样数据和label才能对的上。

sampler那样改一下，真的有效，模型的收敛速度正常了，感谢！
和您的写法比，我原来的写法好像只有`rank in [2,3]`的时候少写了`drop_last=True`，可能是这里导致的
